calculator/output_test.go

```diff
@@ -24,117 +24,129 @@ func init() {
+func TestOutput(t *testing.T) {
+	tc := []testdata{
+		{
+			name:      "Fireball",
+			buildData: "../testdata/builds/Fireball.xml",
+			baseDamage: map[OutTable]map[string]float64{
+				OutTableMainHand: {
+					"TotalMin":      0.9523809523809523,
+					"TotalMax":      2.8571428571428568,
+					"AverageHit":    1.9047619047619047,
+					"AverageDamage": 1.8857142857142855,
+					"TotalDPS":      2.2628571428571425,
+				},
+			},
+			skillDamage: []skillGroup{
+				{
+					name:        "Fireball level 1",
+					socketGroup: 2,
+					damage: map[string]float64{
+						"TotalMin":      9,
+						"TotalMax":      14,
+						"AverageHit":    11.844999999999999,
+						"AverageDamage": 11.845,
+						"TotalDPS":      15.793333333333333,
+					},
+				},
+				{
+					name:        "Fireball level 20",
+					socketGroup: 3,
+					damage: map[string]float64{
+						"TotalMin":      1640,
+						"TotalMax":      2460,
+						"AverageHit":    2111.5,
+						"AverageDamage": 2111.5,
+						"TotalDPS":      2815.333333333333,
+					},
+				},
+				{
+					name:        "Fireball level 1 Added Cold level 1",
+					socketGroup: 4,
+					damage: map[string]float64{
+						"TotalMin":      24,
+						"TotalMax":      36,
+						"AverageHit":    30.9,
+						"AverageDamage": 30.9,
+						"TotalDPS":      41.199999999999996,
+					},
+				},
+				{
+					name:        "Fireball level 20 Added Cold level 1",
+					socketGroup: 5,
+					damage: map[string]float64{
+						"TotalMin":      1655,
+						"TotalMax":      2482,
+						"AverageHit":    2130.555,
+						"AverageDamage": 2130.555,
+						"TotalDPS":      2840.74,
+					},
+				},
+				{
+					name:        "Fireball level 20 Added Cold level 20",
+					socketGroup: 6,
+					damage: map[string]float64{
+						"TotalMin":      2202,
+						"TotalMax":      3304,
+						"AverageHit":    2835.5899999999997,
+						"AverageDamage": 2835.5899999999992,
+						"TotalDPS":      3780.7866666666655,
+					},
+				},
+			},
+		},
+	}
 
-	testza.AssertEqual(t, float64(2202), env.Player.Output["TotalMin"])
-	testza.AssertEqual(t, float64(3304), env.Player.Output["TotalMax"])
-	testza.AssertEqual(t, 2835.5899999999997, env.Player.Output["AverageHit"])
-	testza.AssertEqual(t, 2835.5899999999992, env.Player.Output["AverageDamage"])
-	testza.AssertEqual(t, 3780.7866666666655, env.Player.Output["TotalDPS"])
+	for _, test := range tc {
+		t.Run(test.name, func(t *testing.T) {
+			d, err := os.ReadFile("../testdata/builds/Fireball.xml")
+			if err != nil {
+				t.Fatal(err)
+			}
+			build, err := builds.ParseBuild(d)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Test without skills
+			if test.baseDamage != nil {
+				skills := build.Skills.SkillSets
+				build.Skills.SkillSets = []pob.SkillSet{}
+				env := NewCalculator(*build).BuildOutput(OutputModeMain)
+				for weapon := range test.baseDamage {
+					checkDamage(t, test.baseDamage[weapon], env.Player.OutputTable[weapon])
+				}
```

Comment:
I think this can be further simplified. Create a function
```go
func assertNestedMapEqual[K comparable, V any](t *testing.T, expected, got map[K]V) {
```
that then loops internally and calls `assertMapEqual`.

Then this loop and call would just become
```go
assertNestedMapEqual(t, test.baseDamage, env.Player.OutputTable)
```

Reply:
It needs to be a little more complex than that as you can't pass a V any to a function with a map constraint, but added something that works for single-level nesting. Sadly the current impl of generics doesn't really allow you to do this recursively.